qiskit/transpiler/passes/optimization/inverse_cancellation.py

```diff
@@ -0,0 +1,136 @@
+"""
+from typing import List, Tuple, Union
+
+from qiskit.circuit import Gate
+from qiskit.dagcircuit import DAGCircuit
+from qiskit.transpiler.basepasses import TransformationPass
+from qiskit.transpiler.exceptions import TranspilerError
+
+
+class InverseCancellation(TransformationPass):
+    """Cancel specific Gates which are inverses of each other when they occur back-to-
+    back."""
+
+    def __init__(self, gates_to_cancel: List[Union[Gate, Tuple[Gate, Gate]]]):
+        """Initialize InverseCancellation pass.
+
+        Args:
+            gates_to_cancel: list of gates to cancel
+
+        Raises:
+            TranspilerError:
+                Initalization raises an error when the input is not a self-inverse gate
+                or a two-tuple of inverse gates.
+        """
+
+        for gates in gates_to_cancel:
+            if isinstance(gates, Gate):
+                if gates != gates.inverse():
+                    raise TranspilerError("Gate {} is not self-inverse".format(gates.name))
+            elif isinstance(gates, tuple):
+                if len(gates) != 2:
+                    raise TranspilerError(
+                        "Too many or too few inputs: {}. Only two are allowed.".format(gates)
+                    )
+                elif gates[0] != gates[1].inverse():
+                    raise TranspilerError(
+                        "Gate {} and {} are not inverse.".format(gates[0].name, gates[1].name)
+                    )
+            else:
+                raise TranspilerError(
+                    "InverseCancellation pass does not take input type {}. Input must be"
+                    " a Gate.".format(type(gates))
+                )
+
+        self.gates_to_cancel = gates_to_cancel
+        super().__init__()
+
+        self_inverse_gates = []
+        inverse_gate_pairs = []
+
+        for gates in self.gates_to_cancel:
+            if isinstance(gates, Gate):
+                self_inverse_gates.append(gates)
+            else:
+                inverse_gate_pairs.append(gates)
+
+    def run(self, dag: DAGCircuit):
+        """Run the InverseCancellation pass on `dag`.
+
+        Args:
+            dag: the directed acyclic graph to run on.
+
+        Returns:
+            DAGCircuit: Transformed DAG.
+        """
+
+        dag = self._run_on_self_inverse(dag, self_inverse_gates)
+        return self._run_on_inverse_pairs(dag, inverse_gate_pairs)
+
+    def _run_on_self_inverse(self, dag: DAGCircuit, self_inverse_gates: List[Gate]):
+        """
+        Run self-inverse gates on `dag`.
+
+        Args:
+            dag: the directed acyclic graph to run on.
+            self_inverse_gates: list of gates who cancel themeselves in pairs
+
+        Returns:
+            DAGCircuit: Transformed DAG.
+        """
+        gate_cancel_runs = [dag.collect_runs([gate.name]) for gate in self_inverse_gates]
+        for gate_cancel_run in gate_cancel_runs:
```

Comment:
dag.collect runs should return a list, and now we are putting that into another list, so I think we will probably need to put this for loop into another nested for loop.